stan/math/gpu/opencl_context.hpp

```diff
@@ -0,0 +1,404 @@
+    if (gpu_local_max < local)
+      local = gpu_local_max;
+    snprintf(temp, sizeof(temp), "-D TS=%d -D TS1=%d -D TS2=%d ", local, local,
+             local);
+    std::string kernel_source = "";
+    const char* kernel_group = kernel_info()[kernel_name].group;
+    for (auto kern : kernel_info()) {
+      if (strcmp(kern.second.group, kernel_group) == 0) {
+        kernel_source += kern.second.raw_code;
+      }
+    }
+
+    try {
+      cl::Program::Sources source(
+          1,
+          std::make_pair(kernel_source.c_str(), strlen(kernel_source.c_str())));
+      cl::Program program_ = cl::Program(context(), source);
+      program_.build({device()}, temp);
+
+      cl_int err = CL_SUCCESS;
+      // Iterate over the kernel list and get all the kernels from this group
+      // and mark them as compiled.
+      for (auto kern : kernel_info()) {
+        if (strcmp(kern.second.group, kernel_group) == 0) {
+          kernels()[(kern.first)] = cl::Kernel(program_, kern.first, &err);
+          kern.second.exists = true;
+        }
+      }
+    } catch (const cl::Error& e) {
+      check_ocl_error("Kernel Compilation", e);
+    }
+  }
+
+ public:
+  opencl_context() = default;
+  /**
+   * Returns the kernel specified in kernel_name.
+   * If the kernel has not yet been compiled, the kernel group is compiled
+   * first.
+   *
+   * @brief Passing the name of a kernel compiles all kernels in the same group
+   * as the selected kernel.
+   * OpenCL kernels are compiled JIT, instead of compiling each kernel
+   * individually this function will compile all kernels
+   * in a predefined group. Groupings are made such that kernels commonly
+   * called with one another will be compiled at the same time. For example,
+   * An arithmetic group of kernels compiled together could contain the kernels
+   * for <code> add() </code>, <code> subtract() </code>,
+   * and <code> multiply() </code>. This function will only return the kernel
+   * which was called, but when a user asks for a kernel within the group those
+   * kernels will already be compiled.
+   *
+   * @param[in] kernel_name The kernel name
+   *
+   * @return a copy of the cl::Kernel object
+   */
+  inline cl::Kernel get_kernel(const char* kernel_name) {
+    // Compile the kernel group and return the kernel
+    if (!kernel_info()[kernel_name].exists) {
+      compile_kernel_group(kernel_name);
+    }
+    return kernels()[kernel_name];
+  }
+
+  // FIXME:(Steve/Dan) should probably be deleted before merge
+  void debug(std::ostream& s) {
+    s << "inside opencl_context" << std::endl;
+    s << " * platform_name_: "
+      << opencl_context_base::getInstance().platform_name_ << std::endl;
+  }
+
+  /**
+   * Returns the description of the OpenCL platform and device that is used.
+   * Devices will be a GPU and Platforms are a specific OpenCL implimenation
+   * such as AMD SDK's or Nvidia's OpenCL implimentation.
+   */
+  inline std::string description() const {
```

Comment:
Is it okay to have `std::string` here? Not sure if this conflicts with how we use `char*` everywhere else

Reply:
yes. we shouldn't be using the description method very often.